pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -422,11 +434,76 @@ if [ "$action" = dry-build ]; then
     extraBuildFlags+=(--dry-run)
 fi
 
+if [ "$action" = list-generations ]; then
+    if [ ! -L "$profile" ]; then
+        echo "No profile \`$(basename $profile)' found" >&2
+        exit 1
+    fi
+
+    generation_from_dir() {
+        generation_dir="$1"
+        generation_base="$(basename "$generation_dir")" # Has the format "system-123-link" for generation 123
+        no_link_gen="${generation_base%-link}"  # remove the "-link"
+        echo ${no_link_gen##*-} # remove everything before the last dash
+    }
+    describe_generation(){
+        generation_dir="$1"
+        generation_number="$(generation_from_dir "$generation_dir")"
+        nixos_version="$(cat "$generation_dir/nixos-version" 2> /dev/null || echo "Unknown")"
+
+        kernel_dir="$(dirname "$(realpath "$generation_dir/kernel")")"
+        kernel_version="$(ls "$kernel_dir/lib/modules" || echo Unknown)"
+
+        configurationRevision="$($generation_dir/sw/bin/nixos-version --configurationRevision 2> /dev/null || :)"
+
+        # Old nixos-version output ignored unknown flags and just printed the version
+        # therefore the following workaround is done not to show the default output
+        nixos_version_default="$($generation_dir/sw/bin/nixos-version)"
+        if [ "$configurationRevision" = "$nixos_version_default" ]; then
+             configurationRevision=""
+        fi
+
+        build_date="$(date --date="@$(stat "$generation_dir" --format=%W)" "+%a %F %T")"
+
+        if [ -z "$json" ]; then
+            unset current_generation_tag
+        else
+            current_generation_tag="false"
+        fi
+        if [ "$(basename "$generation_dir")" = "$(readlink $profile)" ]; then
+            if [ -z "$json" ]; then
+                current_generation_tag="  (current)"
+            else
+                current_generation_tag="true"
+            fi
+        fi
+
+        if [ -z $json ]; then
+            echo "$generation_number,$build_date,$nixos_version,$kernel_version,$configurationRevision$current_generation_tag"
+        else
+            echo "\"$generation_number\": { \"date\": \"$build_date\", \"nixosVersion\": \"$nixos_version\", \"kernelVersion\": \"$kernel_version\", \"configurationRevision\": \"$configurationRevision\", \"current\": $current_generation_tag}"
+        fi
+    }
+
+    find $(dirname $profile) -regex "$profile-[0-9]+-link" |
+        sort |
+        while read -r generation_dir; do
+            describe_generation "$generation_dir"
+        done |
+        if [ -z "$json" ]; then
+            column --separator "," --table --table-columns "Generation,Build-date,NixOS version,Kernel,Configuration Revision" |
+                ${PAGER:less}
+        else
+            tr '\n' ',' | sed 's/,$//' | cat <(echo '{') - <(echo '}')
```

Comment:
Should the `--json` output be an array instead of an object? That way it might be easier to get the information for relative generations (`echo '[{"a":"1"},{"a":"2"}]' | jq last`), but as an object it is easier for absolute numbers (`echo '{"1": {"a":"3"},"2":{"a":"4"}}' | jq '."1"'`).

Reply:
I would favor an array, if we make sure, that the generations are ordered correctly (I think currently this is by accident / filesystem ordering, right?), and an object otherwise. But no strong opinion on my part.

Reply:
No, the output of `find` is [`sort`ed](https://github.com/NixOS/nixpkgs/blob/ea6cac67219731b5afddcfc86f9a72b86526db22/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L489). In a previous implementation I used globbing without special sorting, but this didn't play nicely if you had profiles like `test` and `test-1` - you probably still remembered this solution...